BulkData Import Operation: Switch S3 and File to use Byte Specific read locations #2549

prb112 opened the pull request:
- *FILE* Provider
- Convert to use SeekableByteChannel with FileProvider
- Enables skipping to a byte versus read and discard buffer approach
- Use an upper bound read BLOCK to 512K
- Improved Read Performance from 33 Resources per second to 66 per
second in local tests
- Increase Import READ from a block of 20 to 50 (matrixes with block size)

- *S3* Provider
- Switch to retrieve bytes by range
- Added tests to test CountInputStreamReader to test
	- Blocks of 32K buffered while using readLine
	- Multiple Lines
	- Stripping CR from the Input
       - Establish Drain process for S3ObjectInputStream before closing it
	- Implement Range based retrieval in the S3Provider
	- Added a Test Provider Main to suck content down from S3
	- And ImportOperationTest for a decent sized resource and test data (patient-large.ndjson - 2 patient resources)

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

lmsurpre commented:
@prb112 where did [fhir-server-test/src/test/resources/testdata/import-operation/patient-large.ndjson](https://github.com/IBM/FHIR/pull/3384/files#diff-2feff63e2b0f30792517e10497a405d8859378257485dcd4e40ab1810e48631c) come from?
is that a synthea-generated thing or something we came up with?

prb112 commented:
> @prb112 where did patient-large.ndjson come from? is that a synthea-generated thing or something we came up with?

I built it off one of our examples, I needed something larger than 512K

lmsurpre reviewed (approved):
LGTM, though I think it needs substantial testing

prb112 commented:
> LGTM, though I think it needs substantial testing

Absolutely.. it should be run through a regression bucket. Note, it does follow the same pattern as Azure

tbieste commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/jbatch/context/BatchContextAdapter.java:
```diff
@@ -133,6 +133,7 @@ public BulkDataContext getStepContextForImportChunkReader() {
         context(ctx);
```
Update copyright date.

prb112 replied:
I have it in my other open PR. I'm going to defer this, as it'll fix when that is merged.

prb112 replied:
PR https://github.com/IBM/FHIR/pull/3393 is the one that has it

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -41,9 +44,10 @@
 
     private static final Logger logger = Logger.getLogger(FileProvider.class.getName());
 
+    private static final long READ_BLOCK_OPT = 524288L;
```
Any rationale for picking this size? If so, might be worth a comment

prb112 replied:
Yes, I tried this with a few difference sizes.  Let me add an implNote

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
-                    line++;
-                    br.readLine(); // We know the file has at least this number.
-                }
+            try (RandomAccessFile raf = new RandomAccessFile(Paths.get(getFilePath(workItem)).toFile(), "r")) {
```
Random access might cause trouble depending on how the file-system is provided.

prb112 replied:
Yes, that's right, it does speed up the import in this case. The downside of it is actually the same for us.  The prior code used read-to-line which is inefficient.  Some file systems use this approach of scan to location.  I think it's a nominal difference, and in most of the FileProvider usecases it's using the local file system on a developer machine.  

Let me know if there further downstream actions/preferences.

prb112 replied:
I added BufferedInputStream inbetween the Inputstream and Count

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+            int lineLength = 0;
+
+            // Protect against attacks with a max line length (and max size we support in the db).
+            while (read && lineLength < MAX_LENGTH_PER_LINE) {
```
shouldn't there be an IOException if we hit MAX_LENGTH_PER_LINE?

prb112 replied:
I'll make an implnote, I chose not to do that, rather deferring to the downstream invalid parse.

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+                    length++;
+                    // We only need to account for line length here.
+                    lineLength++;
+                    r = this.read();
```
Could be very slow unless the underlying InputStream is buffered.

prb112 replied:
This code is based on the BufferedReader. It follows a similar pattern.

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
 
+    /**
+     * This specialized class enables a BufferedReader.readLine like implementation
+     * without the buffering as there is a local file with direct read and seek.
```
You still need to use buffering to avoid hitting the file-system layer too many times

prb112 replied:
I'll clarify

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+     * This specialized class enables a BufferedReader.readLine like implementation
+     * without the buffering as there is a local file with direct read and seek.
+     */
+    private static class CountInputStreamReader extends InputStreamReader {
```
What about multi-byte characters? I think this needs to be based of a Reader, not an InputStream.

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
+     * This specialized class enables a BufferedReader.readLine like implementation
+     * without the buffering as there is a local file with direct read and seek.
+     */
+    private static class CountInputStreamReader extends InputStreamReader {
```
I think this is broken for multibyte characters. The seek is based on a byte offset on the stream, but you are counting the number of characters read by the reader. The two are not equivalent.

prb112 replied:
Do we really support DBCS? I've tried with unicode 4byte and 3byte and 2byte.  It works well.

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/FileProvider.java:
```diff
@@ -79,43 +83,104 @@ public long getSize(String workItem) throws FHIRException {
-                if (br != null) {
-                    br.close();
+                    // The Channel enables us to skip the future seek.
+                    this.transientUserData.setCurrentBytes(this.transientUserData.getCurrentBytes() + reader.getLength());
```
reader length is the number of characters read, not the number of bytes read

prb112 replied:
I'll add a custom test to show it works properly

prb112 replied:
Added the tests for Files/S3 - looks great in my testing of the stream readers.

punktilious commented on fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:
```diff
@@ -186,11 +203,23 @@ public ListObjectsV2Result getListObject(String prefix, String continuationToken
+     * @return
+     * @throws FHIRException
+     */
+    public long getSize(AmazonS3 c, String bn, String workItem) throws FHIRException {
```
What units for size? Bytes or characters?

prb112 replied:
bytes - added implNote

prb112 commented:
I rewrote the processing of multiple bytes and added a fairly complicated test related to it. 
If the PR passes, I'd probably recommend carefully accepting the PR post next release. 
It requires extensive testing, and I will not be available to support any necessary updates.

lmsurpre commented:
We decided to hold off on this one for now.  Might reopen after our 4.11
